sos/cleaner/mappings/ipv6_map.py

```diff
@@ -0,0 +1,264 @@
+
+import ipaddress
+
+from random import getrandbits
+from sos.cleaner.mappings import SoSMap
+
+
+def generate_hextets(hextets):
+    """Generate a random set of hextets, based on the length of the source
+    hextet. If any hextets are compressed, keep that compression.
+
+    E.G. '::1234:bcd' will generate a leading empty '' hextet, followed by two
+    4-character hextets.
+
+    :param hextets:     The extracted hextets from a source address
+    :type hextets:      ``list``
+
+    :returns:           A set of randomized hextets for use in an obfuscated
+                        address
+    :rtype:             ``list``
+    """
+    return [random_hex(4) if h else '' for h in hextets]
+
+
+def random_hex(length):
+    """Generate a string of size length of random hex characters.
+
+    :param length:  The number of characters to generate
+    :type length:   ``int``
+
+    :returns:       A string of ``length`` hex characters
+    :rtype:         ``str``
+    """
+    return f"{getrandbits(4*length):0{length}x}"
+
+
+class SoSIPv6Map(SoSMap):
+    """Mapping for IPv6 addresses and networks.
+
+    Much like the IP map handles IPv4 addresses, this map is designed to take
+    IPv6 strings and obfuscate them consistently to maintain network topology.
+    To do this, addresses will be manipulated by the ipaddress library.
+
+    If an IPv6 address is encountered without a netmask, it is assumed to be a
+    /64 address.
+    """
+
+    networks = {}
+
+    ignore_matches = [
+        r'^::1/.*',
+        r'::/0',
+        r'fd53:.*',
+        r'^53..:'
+    ]
+
+    compile_regexes = False
+    version = 1
+
+    def conf_update(self, config):
+        """Override the base conf_update() so that we can load the existing
+        networks into ObfuscatedIPv6Network() objects for the current run.
+        """
+        if 'networks' not in config:
+            return
+        for network in config['networks']:
+            _orig = ipaddress.ip_network(network)
+            _obfuscated = config['networks'][network]['obfuscated']
+            _net = self._get_network(_orig, _obfuscated)
+            self.dataset[_net.original_address] = _net.obfuscated_address
+            for host in config['networks'][network]['hosts']:
+                _ob_host = config['networks'][network]['hosts'][host]
+                _net.add_obfuscated_host_address(host, _ob_host)
+                self.dataset[host] = _ob_host
+
+    def sanitize_item(self, ipaddr):
+        _prefix = ipaddr.split('/')[-1] if '/' in ipaddr else ''
+        _ipaddr = ipaddr
+        if not _prefix:
+            # assume a /64 default per protocol
+            _ipaddr += "/64"
+        try:
+            _addr = ipaddress.ip_network(_ipaddr)
+            # ipaddr was an actual network per protocol
+            _net = self._get_network(_addr)
+            _ipaddr = _net.obfuscated_address
+        except ValueError:
+            # host bits are set, first get the network
```

Comment:
Worth expanding the explanation when the exception can be raised? `ipaddress.ip_network(_ipaddr)` raises

    ValueError: 2620:52:0:2d80::4fe/64 has host bits set

but it took me some debugging to understand it..

Reply:
Ack, can do.

Reply:
Added